kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java

```diff
@@ -362,29 +498,50 @@ private void startGCSToBQLoadTask() {
     GCSToBQLoadRunnable loadRunnable = new GCSToBQLoadRunnable(getBigQuery(), bucket);
 
     int intervalSec = config.getInt(BigQuerySinkConfig.BATCH_LOAD_INTERVAL_SEC_CONFIG);
-    gcsLoadExecutor.scheduleAtFixedRate(loadRunnable, intervalSec, intervalSec, TimeUnit.SECONDS);
+    loadExecutor.scheduleAtFixedRate(loadRunnable, intervalSec, intervalSec, TimeUnit.SECONDS);
+  }
+
+  private void maybeStartMergeFlushTask() {
+    long intervalMs = config.getLong(config.MERGE_INTERVAL_MS_CONFIG);
+    if (intervalMs == -1) {
+      logger.info("{} is set to -1; periodic merge flushes are disabled", config.MERGE_INTERVAL_MS_CONFIG);
+      return;
+    }
+    logger.info("Attempting to start upsert/delete load executor");
+    loadExecutor = Executors.newScheduledThreadPool(1);
```

Comment:
Sort of... the executor here just periodically calls `MergeQueries::mergeFlushAll`, which in turn does little more than make a series of calls to `executor::execute`, so the only purpose of this single-threaded executor is to periodically wake up and then schedule the actual merge flushing tasks on a different executor.

See https://github.com/C0urante/kafka-connect-bigquery/blob/6dda50e4707d5df365ef836c3d4b7648097f010f/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/MergeQueries.java#L111

Is there something wrong with this/do you think it can be implemented better somehow?

Reply:
I'm not sure what will happen to the executor instances once the reference to it is assigned to an instance created in another thread. GC may come along and see no references to the instance and decide to GC it without properly shutting it down. Or, it may chug along in the background and you end up with many independently scheduled `mergeFlushAll` calls, each at fixed rate but slightly offset intervals. They likely would not be `shutdown` properly on JVM shutdown or even task rebalances since the reference has been overwritten.

I think it might be better to create the background thread just once on the worker as a singleton instance and have all tasks schedule on the shared instance. Especially since this thread is not doing anything except schedule the `mergeFlushAll` to be executed on another thread. WDYT?

Reply:
Sorry, which executor are you referring to? The `loadExecutor` and `executor` fields are both created and assigned in the same thread that calls `BigQuerySinkTask::start`, and are both cleaned up during `BigQuerySinkTask::stop`. I'm not sure I understand how GC or threading will come into play here?

Using a shared static instance of the executor doesn't seem like a great idea given that there may be task instances of different connectors running on the same worker, but I also don't fully understand the objections to the current model, so there might be something I'm missing here.

Reply:
My understanding is that `BigQuerySinkTask::start` would be called once for each task, each on its own thread. Since the `loadExecutor` field is not thread local, each thread would get access to the shared reference in memory. So if each `start()` method creates a new instance and assigns it to the field, I am not sure what happens to the previous instances created by other threads. I could be misreading the `Worker` initialization code or misunderstanding the threading model entirely though. LMK if you think I am missing something here.

You are right that multiple connectors sharing the global singleton executor instance could be a concern. Rather than have a global singleton, the existing `gcsLoadExecutor` seems to have the connector detect if it is generating config for task zero and explicitly set a flag on the config to tell the task whether it is task zero and needs to create the executor or not. I don't think this will work to create one per worker as the connector doesn't even know what workers the tasks will be scheduled on. You could maybe initialize the executor to null with an AtomicReference to it like you've done for `bigQuery` and `schemaManager`. You'd have to ref count and do executor shutdown when ref count reaches zero though. A bit more involved than a global singleton and a JVM shutdown hook, but it can be done.

Reply:
The `loadExecutor` field is non-static, so each task creates and uses its own. Every thread a task creates will share that executor, but those will still be completely isolated from other tasks. It's similar with the `bigQuery` and `schemaManager` fields; while not thread-local, those are task-local. (And they're atomic references just to avoid potential races when multiple threads from a single task try to initialize them for the first time.)

Reply:
I think the reason the `loadExecutor` field is isolated to a single task when GCS batch loading is enabled isn't to limit sharing of the `Executor` instance itself, but rather, to ensure that there's only one task at a time triggering batch loading jobs in GCS.

Reply:
We discussed offline. Each task would get a new instance and not share the `executor` field with other tasks, so my concern was void. The only real issue I still see here is slightly wasting threads because of a background thread created per task just to schedule the periodic call to `mergeFlushAll`. But, it will be a pain to try to consolidate that to have 1 executor per worker, so we can just leave as is for now and improve it later if this becomes an issue.